build.wake

```diff
@@ -0,0 +1,430 @@
+#
+# Utility Functions
+#
+
+# This expects a sorted list of paths.
+# If a path is a prefix of the path that follows it, it is filtered out.
+def filterDirs sortedFiles =
+  def helper acc list = match list
+    head1, head2, tail =
+      def isPrefixOf path = matches (regExpCat (path.quote, `(/.+)?`, Nil))
+      if isPrefixOf head1 head2
+      then helper acc (head2, tail)
+      else helper (head1, acc) (head2, tail)
+    _ = list ++ acc
+  helper Nil sortedFiles
+  | reverse
+
+def getDestDirs srcDir destDir inputs =
+  def srcDirRegex = regExpCat (`^`, "{srcDir}/..".simplify.quote, `/+`, Nil)
+  inputs
+  | mapFlat (
+    _
+    | getPathName
+    | simplify
+    | replace srcDirRegex ""
+    | extract `(.*/)[^/]*`
+    | map ("{destDir}/{_}")
+  )
+  | distinctBy scmp
+
+def copyDir srcDir destDir fileInputs =
+  def dirInputs =
+    getDestDirs srcDir destDir fileInputs
+    | map mkdir
+  def inputs = dirInputs ++ fileInputs
+  def from_ = simplify "{srcDir}/.."
+  def srcs =
+    fileInputs
+    | map (
+      _
+      | getPathName
+      | from_.relative
+      | simplify
+    )
+    | filter (! matches `\.\./.*` _)
+    | sortBy (_ <~ _)
+    | filterDirs
+  def cmdline =
+    "rsync", "--inplace", "--relative",
+    "--ignore-times", srcs ++ (relative from_ destDir, Nil)
+  makePlan cmdline inputs
+  | setPlanDirectory from_
+  | runJob
+  | getJobOutputs
+
+def filterWakeAndDotFiles =
+  _
+  | filter (!matches `(.*/)*\..*` _.getPathName)
+  | filter (!matches `.*\.wake` _.getPathName)
+
+#
+# Plan Resources
+#
+
+global def defaultFreedomESDKResources =
+  "riscv-tools/2019.08.0",
```

Comment:
Is possible to support a list of tools versions?
And also allow caller to specify what tools version to use, similar to choosing VCS vs Verilator.
By default it can use the version define here.
But caller may be able to select a particular version or all of the support tools versions.

Reply:
I've implemented a fix for this with the FreedomESDKBuildEnv tuple.